doc/man/man1/rmtxop.1
.\" RCSid "$Id$"
.TH RMTXOP 1 7/8/97 RADIANCE
.SH NAME
rmtxop - concatenate, add, multiply, divide, transpose, scale, and convert matrices
.SH SYNOPSIS
.B rmtxop
[
.B \-v
][
.B \-f[afdc]
][
.B \-t
][
.B "\-s sf .."
][
.B "\-c ce .."
]
.B m1
[
.B "+*/"
]
.B ".."
.SH DESCRIPTION
.I Rmtxop
loads and concatenates or adds/multiplies/divides
together component matrix files given on the command line.
Each file must have a header containing the following variables:
.sp
.nf
NROWS={number of rows}
NCOLS={number of columns}
NCOMP={number of components}
FORMAT={ascii|float|double|32-bit_rle_rgbe|32-bit_rle_xyze}
.sp
.fi
The number of components indicates that each matrix element is actually
composed of multiple elements, most commonly an RGB triple.
This is essentially dividing the matrix into planes, where each component
participates in a separate calculation.
If an appropriate header is not present, it may be added with a call to
.I rcollate(1).
A matrix may be read from the standard input using a hyphen by itself ('-')
in the appropriate place on the command line.
.PP
Any of the matrix inputs may be read from a command
instead of a file by
using quotes and a beginning exclamation point ('!').
.PP
Two special cases are handled for component matrices that are either
XML files containing BTDF data, or Radiance picture files.
In the first case, a BSDF library is used to load and interpret the
transmission matrix.
(XML files cannot be read from the standard input or from a command.)\0
In the second case, the RGBE or XYZE values are loaded in a 3-component
matrix where the number of columns match the X-dimension of the picture, and
the number of rows match the Y-dimension.
The picture must be in standard pixel ordering, and the first row
is at the top with the first column on the left.
.PP
Before each file, the
.I \-t
and
.I \-s
or
.I \-c
options may be used to modify the matrix.
The
.I \-t
option transposes the matrix, swapping rows and columns.
The
.I \-s
option applies the given scalar factor(s) to the elements of the matrix.
If only one factor is provided,
it will be used for all components.
If multiple factors are given, their number must match the number of matrix
components.
Alternatively, the
.I \-c
option may be used to "transform" the element values, possibly changing
the number of components in the matrix.
For example, a 3-component matrix can be transformed into a single-component
matrix by using
.I \-c
with three coefficients.
A four-component matrix can be turned into a two-component matrix using 8
coefficients, where the first four coefficients will be used to compute
the first new component, and the second four coefficients
yield the second new component.
Note that the number of coefficients must be an even multiple of the number
of original components.
The
.I \-s
and
.I \-c
options are mutually exclusive, insofar as they cannot be applied together
to the same input matrix.
.PP
If present, the second and subsequent matrices on the command
line are concatenated to the result unless separated by a plus ('+'),
asterisk ('*'), or forward slash ('/') symbol,
in which case the matrix elements are added, multiplied, or divided together,
respectively.
(Note that the asterisk must be quoted or escaped in most shells.)\0
In the case of addition, the two matrices involved must have the same number
of components.
For element-wise multiplication and division, the second matrix is
permitted instead to have a single component per element, which will be
applied equally to all components of the first matrix.
If element-wise division is specified, any zero elements in the second
matrix will result in a warning and the corresponding component(s) in the
first matrix will be set to zero.
.PP
The number of components in the new matrix after applying any
.I -c
transform must agree with the prior result.
For concatenation (matrix multiplication), the number of columns
in the prior result must equal the number of rows in the new matrix, and
the result will have the number of rows of the previous and the number
of columns of the new matrix.
In the case of addition, multiplication, and division,
the number of rows and columns of the prior result and the
new matrix must match, and will not be changed by the operation.
.PP
Results are sent to the standard output.
By default, the values will be written in the lowest resolution format
among the inputs, but the
.I \-f
option may be used to explicitly output components
as ASCII (-fa), binary doubles (-fd), floats (-ff), or RGBE colors (-fc).
In the latter case, the actual matrix dimensions are written in the resolution
string rather than the header.
Also, matrix results written as Radiance pictures must have either one
or three components.
In the one-component case, the output is written as grayscale.
.PP
The
.I \-v
option turns on verbose reporting, which announces each operation.
.SH EXAMPLES
To concatenate two matrix files with a BTDF between them and write
the result as binary double:
.IP "" .2i
rmtxop -fd view.vmx blinds.xml exterior.dmx > dcoef.dmx
.PP
To convert a BTDF matrix into a Radiance picture:
.IP "" .2i
rmtxop -fc blinds.xml > blinds.hdr
.PP
To scale a matrix by 4 and add it to the transpose of another matrix:
.IP "" .2i
rmtxop -s 4 left.mtx + -t right.mtx > result.mtx
.PP
To send the elements of a binary matrix to 
.I rcalc(1)
for further processing:
.IP "" .2i
rmtxop -fa orig.mtx | rcollate -ho -oc 1 | rcalc [operations]
.SH AUTHOR
Greg Ward
.SH "SEE ALSO"
cnt(1), getinfo(1), histo(1), neaten(1), rcalc(1), rcollate(1),
rcontrib(1), rfluxmtx(1), rlam(1), tabfunc(1), total(1), wrapBSDF(1)


src/util/rmtxop.c
#ifndef lint
static const char RCSid[] = "$Id$";
#endif
/*
 * General component matrix operations.
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include "rtio.h"
#include "resolu.h"
#include "rmatrix.h"
#include "platform.h"

#define MAXCOMP		50		/* #components we support */

typedef struct {
	double		sca[MAXCOMP];		/* scalar coefficients */
	int		nsf;			/* number of scalars */
	double		cmat[MAXCOMP*MAXCOMP];	/* component transformation */
	int		clen;			/* number of coefficients */
	int		transpose;		/* do transpose? */
	int		op;			/* '*' or '+' */
} ROPERAT;				/* matrix operation */

int	verbose = 0;			/* verbose reporting? */

static void
op_default(ROPERAT *op)
{
	memset(op, 0, sizeof(ROPERAT));
	op->op = '.';
}

static RMATRIX *
operate(RMATRIX *mleft, ROPERAT *op, const char *fname)
{
	RMATRIX	*mright = rmx_load(fname);
	RMATRIX	*mtmp;
	int	i;

	if (fname == NULL)
		fname = "<stdin>";
	if (mright == NULL) {
		fputs(fname, stderr);
		fputs(": cannot load matrix\n", stderr);
		return(NULL);
	}
	if (op->transpose) {		/* transpose matrix? */
		mtmp = rmx_transpose(mright);
		if (mtmp == NULL) {
			fputs(fname, stderr);
			fputs(": transpose failed\n", stderr);
			rmx_free(mright);
			return(NULL);
		}
		if (verbose) {
			fputs(fname, stderr);
			fputs(": transposed rows and columns\n", stderr);
		}
		rmx_free(mright);
		mright = mtmp;
	}
	if (op->nsf > 0) {		/* apply scalar(s) */
		if (op->clen > 0) {
			fputs("Options -s and -c are exclusive\n", stderr);
			rmx_free(mright);
			return(NULL);
		}
		if (op->nsf == 1) {
			for (i = mright->ncomp; --i; )
				op->sca[i] = op->sca[0];
		} else if (op->nsf != mright->ncomp) {
			fprintf(stderr, "%s: -s must have one or %d factors\n",
					fname, mright->ncomp);
			rmx_free(mright);
			return(NULL);
		}
		if ((mleft == NULL) | (op->op != '+') &&
				!rmx_scale(mright, op->sca)) {
			fputs(fname, stderr);
			fputs(": scalar operation failed\n", stderr);
			rmx_free(mright);
			return(NULL);
		}
		if (verbose) {
			fputs(fname, stderr);
			fputs(": applied scalar (", stderr);
			for (i = 0; i < op->nsf; i++)
				fprintf(stderr, " %f", op->sca[i]);
			fputs(" )\n", stderr);
		}
	}
	if (op->clen > 0) {		/* apply transform */
		if (op->clen % mright->ncomp) {
			fprintf(stderr, "%s: -c must have N x %d coefficients\n",
					fname, mright->ncomp);
			rmx_free(mright);
			return(NULL);
		}
		mtmp = rmx_transform(mright, op->clen/mright->ncomp, op->cmat);
		if (mtmp == NULL) {
			fprintf(stderr, "%s: matrix transform failed\n", fname);
			rmx_free(mright);
			return(NULL);
		}
		if (verbose)
			fprintf(stderr, "%s: applied %d x %d transform\n",
					fname, mtmp->ncomp, mright->ncomp);
		rmx_free(mright);
		mright = mtmp;
	}
	if (mleft == NULL)		/* just one matrix */
		return(mright);
	if (op->op == '.') {		/* concatenate */
		RMATRIX	*mres = rmx_multiply(mleft, mright);
		if (mres == NULL) {
			fputs(fname, stderr);
			if (mleft->ncols != mright->nrows)
				fputs(": mismatched dimensions for multiply\n",
						stderr);
			else
				fputs(": concatenation failed\n", stderr);
			rmx_free(mright);
			return(NULL);
		}
		if (verbose) {
			fputs(fname, stderr);
			fputs(": concatenated matrix\n", stderr);
		}
		rmx_free(mright);
		rmx_free(mleft);
		mleft = mres;
	} else if (op->op == '+') {
		if (!rmx_sum(mleft, mright, op->nsf ? op->sca : (double *)NULL)) {
			fputs(fname, stderr);
			fputs(": matrix sum failed\n", stderr);
			rmx_free(mright);
			return(NULL);
		}
		if (verbose) {
			fputs(fname, stderr);
			fputs(": added in matrix\n", stderr);
		}
		rmx_free(mright);
	} else if ((op->op == '*') | (op->op == '/')) {
		const char *	tnam = (op->op == '/') ?
					"division" : "multiplication";
		errno = 0;
		if (!rmx_elemult(mleft, mright, (op->op == '/'))) {
			fprintf(stderr, "%s: element-wise %s failed\n",
					fname, tnam);
			rmx_free(mright);
			return(NULL);
		}
		if (errno)
			fprintf(stderr,
				"%s: warning - error during element-wise %s\n",
					fname, tnam);
		else if (verbose)
			fprintf(stderr, "%s: element-wise %s\n", fname, tnam);
		rmx_free(mright);
	} else {
		fprintf(stderr, "%s: unknown operation '%c'\n", fname, op->op);
		rmx_free(mright);
		return(NULL);
	}
	return(mleft);
}

static int
get_factors(double da[], int n, char *av[])
{
	int	ac;

	for (ac = 0; ac < n && isflt(av[ac]); ac++)
		da[ac] = atof(av[ac]);
	return(ac);
}

/* Load one or more matrices and operate on them, sending results to stdout */
int
main(int argc, char *argv[])
{
	int	outfmt = DTfromHeader;
	RMATRIX	*mres = NULL;
	ROPERAT	op;
	int	i;
					/* initialize */
	op_default(&op);
					/* get options and arguments */
	for (i = 1; i < argc; i++)
		if (argv[i][0] && !argv[i][1] &&
				strchr("+*/", argv[i][0]) != NULL) {
			op.op = argv[i][0];
		} else if (argv[i][0] != '-' || !argv[i][1]) {
			char	*fname = NULL;	/* load matrix */
			if (argv[i][0] != '-')
				fname = argv[i];
			mres = operate(mres, &op, fname);
			if (mres == NULL) {
				fprintf(stderr, "%s: operation failed on '%s'\n",
						argv[0], argv[i]);
				return(0);
			}
			op_default(&op);	/* reset operator */
		} else {
			int	n = argc-1 - i;
			switch (argv[i][1]) {	/* get option */
			case 'v':
				verbose = !verbose;
				break;
			case 'f':
				switch (argv[i][2]) {
				case 'd':
					outfmt = DTdouble;
					break;
				case 'f':
					outfmt = DTfloat;
					break;
				case 'a':
					outfmt = DTascii;
					break;
				case 'c':
					outfmt = DTrgbe;
					break;
				default:
					goto userr;
				}
				break;
			case 't':
				op.transpose = 1;
				break;
			case 's':
				if (n > MAXCOMP) n = MAXCOMP;
				op.nsf = get_factors(op.sca, n, argv+i+1);
				i += op.nsf;
				break;
			case 'c':
				if (n > MAXCOMP*MAXCOMP) n = MAXCOMP*MAXCOMP;
				op.clen = get_factors(op.cmat, n, argv+i+1);
				i += op.clen;
				break;
			default:
				fprintf(stderr, "%s: unknown operation '%s'\n",
						argv[0], argv[i]);
				goto userr;
			}
		}
	if (mres == NULL)		/* check that we got something */
		goto userr;
					/* write result to stdout */
	if (outfmt == DTfromHeader)
		outfmt = mres->dtype;
	if (outfmt != DTascii)
		SET_FILE_BINARY(stdout);
	newheader("RADIANCE", stdout);
	printargs(argc, argv, stdout);
	if (!rmx_write(mres, outfmt, stdout)) {
		fprintf(stderr, "%s: error writing result matrix\n", argv[0]);
		return(1);
	}
	/* rmx_free(mres); mres = NULL; */
	return(0);
userr:
	fprintf(stderr,
	"Usage: %s [-v][-f[adfc][-t][-s sf .. | -c ce ..] m1 [+*/] .. > mres\n",
			argv[0]);
	return(1);
}
